File: src/main/leetcode/p225/stack/by/queue/MyStack.java
package p225.stack.by.queue;

import java.util.LinkedList;
import java.util.NoSuchElementException;
import java.util.Queue;

/**
 * Created by senyuanwang on 15/6/12.
 */
public class MyStack {
    private Queue<Integer> first = new LinkedList<>();
    private Queue<Integer> second = new LinkedList<>();

    // Push element x onto stack.
    public void push(int x) {
        first.offer(x);
    }

    // Removes the element on top of the stack.
    public void pop() {
        if (first.size() > 0) {
            while (first.size() > 1) {
                second.offer(first.poll());
            }
            first.poll();
        }
        swapQueue();
    }

    // Get the top element.
    public int top() {
        int rt = 0;
        if (first.size() > 0) {
            while (first.size() > 1) {
                second.offer(first.poll());
            }
            rt = first.peek();
        }
//        swapQueue();
        return rt;
    }

    private void swapQueue() {
        Queue<Integer> tmp = first;
        first = second;
        second = tmp;
    }

    // Return whether the stack is empty.
    public boolean empty() {
        return first.isEmpty() && second.isEmpty();
    }

    public static void main(String[] args) {
        MyStack myStack = new MyStack();
        myStack.push(1);
        myStack.push(2);
        myStack.push(3);
        System.out.println(myStack.top());
        myStack.pop();
        System.out.println(myStack.top());
        myStack.pop();
        System.out.println(myStack.top());
        System.out.println(myStack.empty());
        myStack.pop();
        System.out.println(myStack.empty());
    }
}

File: src/main/leetcode/set200/set220/p225/MyStack.scala
package set200.set220.p225

import scala.collection.mutable

class MyStack() {

  /** Initialize your data structure here. */
  val que = mutable.Queue.empty[Int]
  val bak = mutable.Queue.empty[Int]

  /** Push element x onto stack. */
  def push(x: Int) {
    que.enqueue(x)
  }

  /** Removes the element on top of the stack and returns that element. */
  def pop(): Int = {
    while (que.size > 1) {
      bak.enqueue(que.dequeue())
    }
    val res = que.dequeue()

    while (bak.size > 0) {
      que.enqueue(bak.dequeue())
    }

    res
  }

  /** Get the top element. */
  def top(): Int = {
    var res = 0
    while (que.size > 0) {
      res = que.dequeue()
      bak.enqueue(res)
    }

    while (bak.size > 0) {
      que.enqueue(bak.dequeue())
    }

    res
  }

  /** Returns whether the stack is empty. */
  def empty(): Boolean = {
    que.isEmpty
  }

}
